internal/tools/monitor/oap/collector/plugins/exporters/clickhouse/builder/external_metric/external_metrics.go

```diff
@@ -0,0 +1,305 @@
+		}
+		batch := tableBatch[table]
+
+		numFieldKeys := getnumFieldKeysBuf()
+		numFieldValues := getnumFieldValuesBuf()
+		strFieldKeys := getstrFieldKeysBuf()
+		strFieldValues := getstrFieldValuesBuf()
+		tagKeys := gettagKeysBuf()
+		tagValues := gettagValuesBuf()
+
+		pairs := getpairsBuf()
+		for k, v := range data.Tags {
+			pairs = append(pairs, odata.Tag{Key: k, Value: v})
+		}
+		sort.Slice(pairs, func(i, j int) bool {
+			return pairs[i].Key < pairs[j].Key
+		})
+		for i := range pairs {
+			tagKeys = append(tagKeys, pairs[i].Key)
+			tagValues = append(tagValues, pairs[i].Value)
+		}
+		putpairsBuf(pairs)
+
+		for k, v := range data.Fields {
+			switch vv := v.(type) {
+			case string:
+				strFieldKeys = append(strFieldKeys, k)
+				strFieldValues = append(strFieldValues, vv)
+			default:
+				vfloat64, err := typeconvert.ToFloat64(v)
+				if err != nil {
+					continue
+				}
+				numFieldKeys = append(numFieldKeys, k)
+				numFieldValues = append(numFieldValues, vfloat64)
+			}
+		}
+		// ignore empty
+		if len(numFieldKeys) == 0 && len(strFieldKeys) == 0 {
+			putnumFieldKeysBuf(numFieldKeys)
+			putnumFieldValuesBuf(numFieldValues)
+			putstrFieldKeysBuf(strFieldKeys)
+			putstrFieldValuesBuf(strFieldValues)
+			puttagKeysBuf(tagKeys)
+			puttagValuesBuf(tagValues)
+			continue
+		}
+
+		err1 := batch.AppendStruct(&metric.TableMetrics{
+			OrgName:           data.OrgName,
+			TenantId:          data.Tags[bu.cfg.TenantIdKey],
+			MetricGroup:       data.Name,
+			Timestamp:         time.Unix(0, data.Timestamp),
+			NumberFieldKeys:   numFieldKeys,
+			NumberFieldValues: numFieldValues,
+			StringFieldKeys:   strFieldKeys,
+			StringFieldValues: strFieldValues,
+			TagKeys:           tagKeys,
+			TagValues:         tagValues,
+		})
+		putnumFieldKeysBuf(numFieldKeys)
+		putnumFieldValuesBuf(numFieldValues)
+		putstrFieldKeysBuf(strFieldKeys)
+		putstrFieldValuesBuf(strFieldValues)
+		puttagKeysBuf(tagKeys)
+		puttagValuesBuf(tagValues)
+		if err1 != nil {
+			_ = batch.Abort()
+			return nil, fmt.Errorf("batch append: %w", err1)
+		}
+	}
+	batches := make([]driver.Batch, 0, len(tableBatch))
+	for _, d := range tableBatch {
+		batches = append(batches, d)
+	}
+	return batches, nil
+}
+
+func (bu *Builder) getOrCreateTenantTable(ctx context.Context, data *metric.Metric) (string, error) {
+	key := bu.Retention.GetConfigKey(data.Name, data.Tags)
+	ttl := bu.Retention.GetTTL(key)
+	var (
+		wait  <-chan error
+		table string
+	)
+	if len(key) > 0 {
+		wait, table = bu.Creator.Ensure(ctx, data.Tags[lib.OrgNameKey], key, tablepkg.FormatTTLToDays(ttl))
+	} else {
+		wait, table = bu.Creator.Ensure(ctx, data.Tags[lib.OrgNameKey], "", tablepkg.FormatTTLToDays(ttl))
+	}
+	if wait != nil {
+		select {
+		case <-wait: // ignore
+		case <-ctx.Done():
+			return table, nil
+		}
+	}
+	return table, nil
+}
+
+var (
+	numFieldKeysBuf, numFieldValuesBuf, strFieldKeysBuf, strFieldValuesBuf, tagKeysBuf, tagValuesBuf sync.Pool
```

Comment:
It's package-level variables. What will happen if we have multiple instances of this provider?

Reply:
metrics will be confused🤪